packages/web/src/common/store/premiumContent/sagas.ts

```diff
@@ -0,0 +1,266 @@
+import {
+  cacheActions,
+  accountSelectors,
+  cacheTracksSelectors,
+  Collectible,
+  User,
+  Track,
+  getContext,
+  Kind,
+  PremiumContentSignature,
+  ID,
+  Chain,
+  FeatureFlags
+} from '@audius/common'
+import { takeLatest, select, call, put } from 'typed-redux-saga'
+
+import {
+  ethNFTsFetched,
+  ETH_NFTS_FETCHED,
+  solNFTsFetched,
+  SOL_NFTS_FETCHED,
+  updatePremiumContentSignatures
+} from 'common/store/premiumContent/actions'
+import { getPremiumTrackSignatureMap } from 'common/store/premiumContent/selectors'
+
+const { getAccountUser } = accountSelectors
+const { getTracks } = cacheTracksSelectors
+
+function hasNotFetchedAllNFTs(account: User) {
+  const { collectibleList, solanaCollectibleList } = account
+  const hasCollectibles = account?.has_collectibles ?? false
+  return (
+    collectibleList === undefined ||
+    solanaCollectibleList === undefined ||
+    (hasCollectibles && !account.collectibles)
+  )
+}
+
+// Map of track ids for which to attempt to get signature.
+// Include token ids for ERC1155 nfts.
+function* getTokenIdMap({
+  tracks,
+  ethCollectibles,
+  solCollectibles,
+  premiumTrackSignatureIdSet
+}: {
+  tracks: { [id: number]: Track }
+  ethCollectibles: Collectible[]
+  solCollectibles: Collectible[]
+  premiumTrackSignatureIdSet: Set<number>
+}) {
+  const trackMap: { [id: number]: string[] } = {}
+
+  // Handle tracks gated by ethereum nft collections
+
+  // Build map of eth nft contract address -> list of token ids
+  // ERC1155 requires token ids to get the balance of a wallet for a given collection
+  // We can ignore token ids for ERC721 nfts
+  const ethContractMap: { [address: string]: string[] } = {}
+  ethCollectibles.forEach((c: Collectible) => {
+    if (!c.assetContractAddress) return
+
+    if (ethContractMap[c.assetContractAddress]) {
+      ethContractMap[c.assetContractAddress].push(c.tokenId)
+    } else {
+      ethContractMap[c.assetContractAddress] = [c.tokenId]
+    }
+  })
+
+  // todo: Handle tracks gated by solana nft collections
+
+  Object.keys(tracks)
+    .map(Number)
+    .filter(Boolean)
+    .forEach((trackId) => {
+      // skip this track entry if we already have a signature for it
+      if (premiumTrackSignatureIdSet.has(trackId)) return
+
+      // skip this track entry if it is not premium or if it is not gated on an nft collection
+      const { is_premium: isPremium, premium_conditions: premiumConditions } =
+        tracks[trackId]
+      if (!isPremium || !premiumConditions || !premiumConditions.nft_collection)
+        return
+
+      // Set the token ids for ERC1155 nfts as the balanceOf contract method
+      // which will be used to determine ownership requires the user's
+      // wallet address and token ids
+
+      // const { nft_collection: nftCollection } = premiumConditions
+      // todo: fix the string nft_collection to be an object
+      // temporarily parse it into object here for now
+      const nftCollection = JSON.parse(
```

Comment:
this is a weird thing that i'll remove later. have to do this right now because the nft_collection is somehow coming back as a string. will fix